compiler/src/dotty/tools/dotc/typer/Checking.scala

```diff
@@ -804,20 +804,21 @@ object Checking {
    *
    */
   def checkAndAdaptExperimentalImports(trees: List[Tree])(using Context): Unit =
-    def nonExperimentalTopLevelDefs(pack: Symbol): Iterator[Symbol] =
-      def isNonExperimentalTopLevelDefinition(sym: Symbol) =
-        sym.isDefinedInCurrentRun
-        && sym.source == ctx.compilationUnit.source
-        && !sym.isConstructor // not constructor of package object
-        && !sym.is(Package) && !sym.name.isPackageObjectName
-        && !sym.isExperimental
-
-      pack.info.decls.toList.iterator.flatMap: sym =>
-        if sym.isClass && (sym.is(Package) || sym.isPackageObject) then
-          nonExperimentalTopLevelDefs(sym)
-        else if isNonExperimentalTopLevelDefinition(sym) then
-          sym :: Nil
-        else Nil
+    def nonExperimentalTopLevelDefs(): List[Symbol] =
+      new TreeAccumulator[List[Symbol]] {
+        override def apply(x: List[Symbol], tree: tpd.Tree)(using Context): List[Symbol] =
+          def addIfNotExperimental(sym: Symbol) =
+            if !sym.isExperimental then sym :: x
+            else x
+          tree match {
+            case tpd.PackageDef(_, contents) => apply(x, contents)
+            case typeDef @ tpd.TypeDef(_, temp: Template) if typeDef.symbol.isPackageObject =>
+              apply(x, temp.body)
+            case vdef: tpd.ValDef if vdef.symbol.isPackageObject => x // ValDef(new PackageObject)
```

Comment:
@smarter Just to confirm - we want to annotate those ValDefs containing package objects then? I'm hesitant since I don't think that is possible to do by annotating by hand with @experimental (it seems we can't annotate package objects at all), but on the other hand this is indeed a top-level def, which is why I'm asking

Reply:
You can annotate a packable object defined as ``object `package` `` but for those generated for top-level defs there isn't a way to do it indeed. For those it might not matter since you cannot extend them, but it seems more uniform to treat everything the same way and therefore to also annotate them here.